toolchain/lex/tokenized_buffer.cpp

```diff
@@ -464,77 +464,106 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
+      emitter_.Emit(source_text.begin(), UnmatchedClosing);
+      Token token = buffer_->AddToken({.kind = TokenKind::Error,
+                                       .token_line = current_line_,
+                                       .column = current_column_,
+                                       .error_length = 1});
+      ++current_column_;
+      source_text = source_text.drop_front();
       return token;
+    };
+
+    // If we have no open groups, this is an error.
+    if (LLVM_UNLIKELY(open_groups_.empty())) {
+      return unmatched_error();
     }
 
-    // Only closing symbols need further special handling.
-    if (!kind.is_closing_symbol()) {
-      return token;
+    Token opening_token = open_groups_.back();
+    // Close any invalid open groups first.
+    if (LLVM_UNLIKELY(buffer_->GetTokenInfo(opening_token).kind !=
+                      kind.opening_symbol())) {
+      CloseInvalidOpenGroups(kind);
+      // This may exhaust the open groups so re-check and re-error if needed.
+      if (open_groups_.empty()) {
+        return unmatched_error();
+      }
+      opening_token = open_groups_.back();
+      CARBON_DCHECK(buffer_->GetTokenInfo(opening_token).kind ==
+                    kind.opening_symbol());
     }
+    open_groups_.pop_back();
 
-    TokenInfo& closing_token_info = buffer_->GetTokenInfo(token);
+    // Now that the groups are all matched up, lex the actual token.
+    Token token = LexOneCharSymbolToken(source_text, kind);
 
-    // Check that there is a matching opening symbol before we consume this as
-    // a closing symbol.
-    if (open_groups_.empty()) {
-      closing_token_info.kind = TokenKind::Error;
-      closing_token_info.error_length = kind.fixed_spelling().size();
+    // Note that it is important to get fresh token infos here as lexing the
+    // open token would invalidate any pointers.
+    buffer_->GetTokenInfo(opening_token).closing_token = token;
+    buffer_->GetTokenInfo(token).opening_token = opening_token;
 
-      CARBON_DIAGNOSTIC(
-          UnmatchedClosing, Error,
-          "Closing symbol without a corresponding opening symbol.");
-      emitter_.Emit(location, UnmatchedClosing);
-      // Note that this still returns true as we do consume a symbol.
-      return token;
+    return token;
+  }
+
+  auto LexSymbolToken(llvm::StringRef& source_text) -> LexResult {
+    // One character symbols and grouping symbols are handled with dedicated
+    // dispatch. We only lex the multi-character tokens here.
+    TokenKind kind = llvm::StringSwitch<TokenKind>(source_text)
+#define CARBON_SYMBOL_TOKEN(Name, Spelling) \
+  .StartsWith(Spelling, TokenKind::Name)
+#define CARBON_ONE_CHAR_SYMBOL_TOKEN(TokenName, Spelling)
+#define CARBON_OPENING_GROUP_SYMBOL_TOKEN(TokenName, Spelling, ClosingName)
+#define CARBON_CLOSING_GROUP_SYMBOL_TOKEN(TokenName, Spelling, OpeningName)
+#include "toolchain/lex/token_kind.def"
+                         .Default(TokenKind::Error);
```

Comment:
There's no more template dispatch after this.

Do you mean using a table-based parsing? It's a bit trickier because we have both 2 and 3 byte symbols to map here. We could conceivably build a 2-byte table for all those cases. But the 3-byte case is going to be a lot.

And i'm not sure we'll even have unique 2-byte prefixes: `<<` vs. `<<=` for example.

One thing that I should maybe clarify in the comments -- this routine still handles most one-character symbols -- any where we might max-munch combine them. The only one-char symbols we handle separately are the ones that we *guarantee* to never fuse into a larger symbol, which is grouping symbols, `,`, and `;`. Those happen to be very *important* cases, but it is a much more narrow optimization than it appears on the surface. For example, even `.` isn't included in case we want to build `.?` or similar in the future.